examples/amm/lib.rs

```diff
@@ -0,0 +1,373 @@
+        lp_shares: Mapping<AccountId, Balance>,
+        fee: Balance,
+    }
+
+    #[ink(impl)]
+    impl Amm {
+        fn has_valid_amount(
+            &self,
+            balance: &Mapping<AccountId, Balance>,
+            _qty: Balance,
+        ) -> Result<(), Error> {
+            let caller = self.env().caller();
+            let shares = balance.get(&caller).unwrap_or(0);
+
+            match _qty {
+                0 => Err(Error::ZeroAmount),
+                _ if _qty > shares => Err(Error::InsufficientAmount),
+                _ => Ok(()),
+            }
+        }
+
+        fn constant_k(&self) -> Balance {
+            self.total_token1 * self.total_token2
+        }
+
+        fn has_active_pool(&self) -> Result<(), Error> {
+            match self.constant_k() {
+                0 => Err(Error::ZeroLiquidity),
+                _ => Ok(()),
+            }
+        }
+
+        #[ink(constructor)]
+        pub fn new(_fee: Balance) -> Self {
+            ink_lang::utils::initialize_contract(Self::new_init);
+
+            Self {
+                fee: if _fee >= 1000 { 0 } else { _fee },
+                ..Default::default()
+            }
+        }
+
+        fn new_init(&mut self) {}
+
+        #[ink(message)]
+        pub fn faucet(&mut self, amount_token1: Balance, amount_token2: Balance) {
+            let caller = self.env().caller();
+            let token1 = self.token1_balance.get(&caller).unwrap_or(0);
+            let token2 = self.token2_balance.get(&caller).unwrap_or(0);
+
+            self.token1_balance
+                .insert(&caller, &(token1 + amount_token1));
+            self.token2_balance
+                .insert(&caller, &(token2 + amount_token2));
+        }
+
+        /// Returns the balances of the user
+        #[ink(message)]
+        pub fn balances(&self) -> (Balance, Balance, Balance) {
+            let caller = self.env().caller();
+            let t1 = self.token1_balance.get(&caller).unwrap_or(0);
+            let t2 = self.token2_balance.get(&caller).unwrap_or(0);
+            let my_shares = self.lp_shares.get(&caller).unwrap_or(0);
+            (t1, t2, my_shares)
+        }
+
+        /// Returns the amount of tokens locked in the pool,total shares issued & trading fee param
+        #[ink(message)]
+        pub fn pool_details(&self) -> (Balance, Balance, Balance, Balance) {
+            (
+                self.total_token1,
+                self.total_token2,
+                self.total_shares,
+                self.fee,
+            )
+        }
+
+        #[ink(message)]
+        pub fn provide_liquidity(
+            &mut self,
+            amount_token1: Balance,
+            amount_token2: Balance,
+        ) -> Result<Balance, Error> {
+            self.has_valid_amount(&self.token1_balance, amount_token1)
+                .unwrap();
+            self.has_valid_amount(&self.token2_balance, amount_token2)
+                .unwrap();
```

Comment:
Make sure to handle these errors correctly instead of unwrapping. Ideally a `Result` is returned from the function, this lets callers potentially handle any error.

No `unwrap`s should be left in the example by the end